llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,461 @@
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI, VRM))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Op.getReg().isVirtual())
+          continue;
+        if (VRM.hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI.isVecOrAccRegClass(MRI.getRegClass(Reg)))
+          continue;
+        // // Subregisters can reuse Constants in the high part of the
+        // // register, therefore renaming a reused physical register can
+        // // prohibit the elimination of a move of constants, thus degrading
+        // // performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse, VRM))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs, VRM, MRI)) {
+          MCRegister PreSwapPhyReg = VRM.getPhys(Reg).id();
+          bool Replaced =
+              replaceReg(Reg, ForbiddenPhysRegs, MF, VRM, LRM, MRI, LIS, TRI);
+
+          if (Replaced) {
+            LLVM_DEBUG(dbgs() << MI);
+            Modified = true;
+            ReplacementRegs.push_back(
+                std::make_pair(PreSwapPhyReg, VRM.getPhys(Reg).id()));
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM.getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
+      }
+    }
+    // dump Reassignment
+    LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+               dbgs() << "\n");
+    for (std::pair<MCRegister, MCRegister> &RegPair : ReplacementRegs) {
+      LLVM_DEBUG(dbgs() << printReg(RegPair.first, &TRI, 0, &MRI) << " -> "
+                        << printReg(RegPair.second, &TRI, 0, &MRI) << "\n");
+    }
+    LLVM_DEBUG(dbgs() << "\n");
+  }
+  return Modified;
+}
+
+static bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                           SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                           SmallVector<MCPhysReg, 16> &UsedVirtRegs,
+                           VirtRegMap &VRM, const MachineRegisterInfo &MRI) {
+  bool FoundPhyReg = UsedPhyRegs.end() != find(UsedPhyRegs, VRM.getPhys(Reg));
+  bool FoundVirtReg =
+      UsedVirtRegs.end() == find(UsedVirtRegs, Register::virtReg2Index(Reg));
+
+  return MRI.getRegClass(Reg) && FoundPhyReg && FoundVirtReg && !MI.isCopy();
+}
+
+static SmallSet<MCPhysReg, 16>
+getBBForbiddenReplacements(const MachineBasicBlock *MBB, const VirtRegMap &VRM,
+                           const MachineRegisterInfo &MRI,
+                           const AIEBaseRegisterInfo &TRI) {
+  SmallSet<MCPhysReg, 16> BlockedPhysRegs;
+  MCPhysReg PhysReg;
+  for (auto &MI : *MBB) {
```

Comment:
As we are iterating with an explicit type in different parts of this code, I think we should expand this auto as well to be consistent..

Reply:
fixed